Remove remaining Kerberos v4 vestiges

frozencemetery opened the pull request:
(For review purposes, it's probably best to review one commit at a time, since the doxygen removal commit is quite large.)

greghudson commented:
ovsec dump support isn't related to v4.

I'd want to check with OpenAFS (@kaduk) before removing the principal conversion functions.

frozencemetery commented:
Apologies, that snuck in from the other branch.  Updated (and fixed tests).

kaduk commented:
It looks like openafs could probably cope with the removal of the 524/425 conv_principal and convert_creds functions; while we do use them even in current  master, there are configure checks and a fallback.  (The functions seem to only be used in 'aklog', the userspace tool for getting a service ticket and massaging it into the expected shape -- the actual core server functionality is implemented independently based on some Heimdal code.)

There are still some cases where the afs3 salt is used, though, and would require operator intervention (not just software updates) to get past.  That said, given that the better alternative has been around for 5.5 years (http://www.openafs.org/pages/security/OPENAFS-SA-2013-003.txt), I won't put a stake in the ground that afs3 salt has to stay in the latest krb5 version indefinitely.

greghudson commented:
Oh, yeah, the afs3 salt type isn't a krb4 feature.  The v4 salt type sort of is and sort of isn't; it's a krb5 feature to facilitate converting databases from v4 to v5.  I would rather treat the removal of the v4 salt type as a DES-removal change than a v4-vestiges-removal cahnge.

greghudson commented:
"Remove srvtab support": doc/basic/keytab_def.rst needs two mentions of srvtabs removed.  Not sure if we care, but I also see some non-user-visible source references to srvtab in lib/krb5/krb/get_in_tkt.c, both the kadm5 and main dejagnu tests, and kprop and kpropd.

"Remove non-functional v4 configuration in krb5.conf": I think the premise of this commit is incorrect.  The v4_realm and v4_instance_convert relations are still used by the conv_princ.c functions.  This stuff can be removed as part of the next commit assuming we do that.

"Remove remaining support for 425 and 524 conversion": I don't believe we can fully remove a public API function without bumping the soname, so we need to leave behind stubs which return errors.

"Remove kadmin RPC support for setting v4 principal": the commit summary should probably say "key", not "principal".  We should bump the sonames of libkadm5clnt and libkadm5srv here.  There's a remaining entry for SETV4KEY_PRINCIPAL in ovsec_kadmd.c:log_badverf().  In kadm5/kadm_rpc.h, we should probably leave behind a comment that 17 was SETV4KEY_PRINCIPAL.

"Remove the v4 and afs3 (v4 only) salttypes": let's defer this commit (both parts) until we're ready to get rid of DES.  In theory people could have converted from a krb4 database long ago and still have DES keys lying around using empty salts.  (Also in theory, people could have specified the v4 salt type combined with a more modern enctype, out of confusion.  We should figure out what happens when there are existing keys whose salt types have been removed; if it's reasonably graceful we can still get rid of the salt type.)

kaduk commented:
> "Remove remaining support for 425 and 524 conversion": I don't believe we can fully remove a public API function without bumping the soname, so we need to leave behind stubs which return errors.

FWIW IIRC the current openafs configure test is not smart enough to discover that these functions exist but are useless stubs.

frozencemetery commented:
- Updated srvtab commit.  I don't see the reference in get_in_tkt.c, but I removed the others.
- Merged v4_realm / v4_instance convert into the 524/425 commit.
- Updated conversion to leave behind stubs.
- Updated kadmin rpc support commit and message to complete removal / breadcrumbs.
- Dropped salttypes from this branch.

@kaduk I'm happy to write openafs a patch if that would help.  If there's fallback it seems easiest to just always use that rather than checking?

frozencemetery commented:
Fair enough.  Dropped that commit.

frozencemetery commented on src/ccapi/test/test_ccapi_ccache.c:
```diff
@@ -303,54 +303,13 @@ int check_cc_ccache_get_credentials_version(void) {
-		failure_count++;
-	}
-
-	// try it with added v5 creds
```
The test here looks (to my eye) like: create a v4 ccache, check the version, set a v5 principal, and then check that the version is _v4_v5.  We definitely can just create a ccache with v5 and then check the version (will be _v5, not _v4_v5), but that's done elsewhere in `check_cc_ccache_get_credentials_version()`.

greghudson replied:
Okay, that's fair.
